### PR TITLE
Fixing Tree Move Bugs #2

### PR DESCRIPTION
**Pull request by gordonkoehn**
This PR addresses 
- #83 

During the development of 
- #82 
some trees generated in the mcmc chain of an initial deep tree could again not be converted to `TreeNodes`, no root was there. The tree was broken. 

There must be a bug in the mcmc moves. After already having fixed one in 
- #48 
and running chains of hundreds of samples it is surprising to see this. Glad to find the bugs now. 
It appears small deep trees test the extreme cases of the tree moves.

So far I have found 2 bugs. One in 
- `_prune_and_reattach_proposal` : the pruned node itself was included in the potential nodes to reattach to
-`_swap_node_labels_move` : confused node levels with indices

I have lost trust in these moves, I have already added test functions, but it is about time to test these moves thoroughly. I'll add tests as proposed in, i.e. this PR also closes
- #14

### CODE REVIEW

File: src/pyggdrasil/tree_inference/_tree.py [hunk 1]
```diff
@@ -68,6 +69,14 @@ def print_topo(self):
...
         return self.to_TreeNode().print_topo()
 
+    def __str__(self):
+        labels = self.labels
+        assert labels is not None
+        df = pd.DataFrame(
+            self.tree_topology.astype(int), labels, labels  # type: ignore
+        )
+        return df.__str__()
+
```
Comment (gordonkoehn) [inline src/pyggdrasil/tree_inference/_tree.py:79]: Easier Debugging

**Review by gordonkoehn — COMMENTED** [review]
Comments on bug fix 1 and 2

File: tests/tree_inference/test_mcmc.py [hunk 2]
```diff
@@ -349,3 +350,131 @@ def test_swap_subtrees_move_fig17_nested_subtrees(seed):
...
+        """Log prob function for testing. - dummy function"""
+        return random.uniform(rng, shape=()).__float__()
+
+    # Run the kernel
+    for i in range(n_moves):
+        rng, rng_now = random.split(rng)
+        tree, _ = mcmc._mcmc_kernel(
+            rng_now, tree, move_probs, log_prob_fn  # type: ignore
+        )
+        assert tr.is_valid_tree(tree)
```
Comment (gordonkoehn) [inline tests/tree_inference/test_mcmc.py:480]: Note this test - runs a mock mcmc with a few samples and checks that the tree is valid still. 
This is the best real test I can think of.

File: tests/tree_inference/test_tree.py [hunk 3]
```diff
@@ -221,3 +225,32 @@ def test_is_not_same_tree():
...
+
+
+@pytest.mark.parametrize("n", [5, 10])
+@pytest.mark.parametrize("seed", [34, 424])
+@pytest.mark.parametrize("tree_type", ["r", "s", "d"])
+def test_is_valid_tree(n: int, seed: int, tree_type: str):
+    """Test is_valid_tree function."""
+
+    tree = make_tree(n, seed, tree_type)
+    assert tr.is_valid_tree(tree) is True
```
Comment (gordonkoehn) [inline tests/tree_inference/test_tree.py:256]: may look circular - yet I am sure that the tree generator functions are working fine

File: src/pyggdrasil/tree_inference/_mcmc.py [hunk 4]
```diff
@@ -79,10 +79,23 @@ def _prune_and_reattach_proposal(key: JAXRandomKey, tree: Tree) -> Tuple[Tree, f
...
+
+    # TODO (Gordon) : consider to remove if checks passed
+    logger.debug("MCMC: Prune and reattach move - pruned node %s", pruned_node)
+
     # get descendants of pruned node
-    descendants = get_descendants(tree.tree_topology, tree.labels, pruned_node)
-    # possible nodes to reattach to - including pruned node for aperiodic case
+    descendants = get_descendants(
+        tree.tree_topology, tree.labels, pruned_node, include_parent=True
+    )
```
[the remark this answers is not in the document]
Comment (pawel-czyz) [reply src/pyggdrasil/tree_inference/_mcmc.py:89]: Wow, this was tricky. Good you found it out!

File: src/pyggdrasil/tree_inference/_mcmc.py [hunk 5]
```diff
@@ -96,15 +109,16 @@ def _prune_and_reattach_proposal(key: JAXRandomKey, tree: Tree) -> Tuple[Tree, f
...
+    node2_idx = jnp.where(tree.labels == node2)[0]
 
     # Copy all the labels
     new_labels = tree.labels
     # Assign label2 to node1...
-    new_labels = new_labels.at[node1].set(label2)
+    new_labels = new_labels.at[node1_idx].set(node2)
     # ... and label1 to node2
-    new_labels = new_labels.at[node2].set(label1)
+    new_labels = new_labels.at[node2_idx].set(node1)
```
[the remark this answers is not in the document]
Comment (pawel-czyz) [reply src/pyggdrasil/tree_inference/_mcmc.py:121]: Right. It may be good to check for this sneaky bug also in the other two moves 
 (permuting the labels and swapping subtrees).

File: src/pyggdrasil/tree_inference/_tree.py [hunk 6]
```diff
@@ -310,3 +321,31 @@ def is_same_tree(tree1: Tree, tree2: Tree) -> bool:
         ) and jnp.all(tree1.labels == tree2_reordered.labels)
 
     return bool(result)
+
+
+def is_valid_tree(tree: Tree) -> bool:
```
Comment (pawel-czyz) [inline src/pyggdrasil/tree_inference/_tree.py:326]: Good idea, it should catch some bugs. Do we know that `to_TreeNode` explicitly check for these four things or we hope it does it implicitly?

File: tests/tree_inference/test_mcmc.py (hunk 2 above)
Comment (pawel-czyz) [reply tests/tree_inference/test_mcmc.py:480]: It's a good one :slightly_smiling_face:

**Review by pawel-czyz — APPROVED** [review]
Very nice you found it! I'd suggest to check for the label/label index bug also in other tree moves, as it's a very sneaky one...

File: src/pyggdrasil/tree_inference/_tree.py (hunk 6 above)
Comment (gordonkoehn) [reply src/pyggdrasil/tree_inference/_tree.py:326]: Yes, converting to `TreeNode` from `Tree` is what so far has caught all the broken trees. To convert it needs to check for the 1) root, 2) self-loops would lead to recursion failing 3)  not connected components leads to multiple roots, also caught

File: src/pyggdrasil/tree_inference/_mcmc.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (gordonkoehn) [reply src/pyggdrasil/tree_inference/_mcmc.py:121]: Yes, they are tested now. 

`swap_subtrees` is correct since #49 

and prune and reattach was also fixed here.
